adapta/storage/blob/s3_storage_client.py

```diff
@@ -132,15 +159,85 @@ def download_blobs(
+        :return:
+        """
+        s3_path = cast_path(blob_path)
+        blobs = self._s3_resource.Bucket(s3_path.bucket).objects.filter(Prefix=s3_path.path)
+        for blob in blobs:
+            if filter_predicate is None or filter_predicate(blob):
+                local_file_path = f"{local_path}/{blob.key}"
+                os.makedirs(os.path.dirname(local_file_path), exist_ok=True)
+                try:
+                    self._s3_resource.meta.client.download_file(s3_path.bucket, blob.key, f"{local_path}/{blob.key}")
+                except StorageClientError as error:
+                    raise RuntimeError(f"Error downloading blob: {error}") from error
+
+    def copy_blob(self, blob_path: DataPath, target_blob_path: DataPath, doze_period_ms: int = 0) -> None:
         """
-        raise NotImplementedError("Not yet implemented in S3StorageClient")
+        Copies a blob from one location to another in S3 storage.
 
-    def copy_blob(self, blob_path: DataPath, target_blob_path: DataPath, doze_period_ms: int) -> None:
+        :param blob_path: Path to the source blob
+        :param target_blob_path: Path to the target location
+        :param doze_period_ms: Not utilized for AWS operations, iIt is included to ensure interface compliance.
         """
-        Not implemented in S3 Client
+        source_s3_path = cast_path(blob_path)
+        target_s3_path = cast_path(target_blob_path)
+
+        source_objects = self._s3_resource.Bucket(source_s3_path.bucket).objects.filter(Prefix=source_s3_path.path)
+
+        for source_object in source_objects:
+            # If the source path is a directory, construct the target object key
+            # If the source path is a file, the target object key is the target path
+            if source_s3_path.path == source_object.key:
+                target_object_key = target_s3_path.path
+            else:
+                target_object_key = source_object.key.replace(source_s3_path.path, target_s3_path.path, 1)
+
+            copy_source = {"Bucket": source_s3_path.bucket, "Key": source_object.key}
+
+            try:
+                self._s3_resource.meta.client.copy(copy_source, target_s3_path.bucket, target_object_key)
+                self._s3_resource.meta.client.head_object(Bucket=target_s3_path.bucket, Key=target_object_key)
+            except StorageClientError as error:
+                print(f"Error copying object: {error}")
+
+    def save_data(self, source_file_path: str, target_file_path: DataPath, doze_period_ms: int = 0) -> None:
         """
-        raise NotImplementedError("Not implemented in S3StorageClient")
+        Saves a target file or folder at `source_file_path` to `file_path`
+
+        :param source_file_path: Source file or folder path.
+        :param target_file_path: Target file path in DataPath notation.
+        :param doze_period_ms: Not utilized for AWS operations, iIt is included to ensure interface compliance.
```

Comment:
```suggestion
        :param doze_period_ms: Not utilized for AWS operations, it is included to ensure interface compliance.
```

Reply:
Another nice catch! Thank you for pointing it out 🙂